bigtable_input.py

```diff
@@ -82,26 +129,76 @@ def get_unparsed_moves_from_last_n_games(n,
         raise ValueError('Cannot find a latest game in the table')
 
     start = int(max(0, latest_game - n))
-    start_row, end_row = get_game_range_row_names(start, latest_game)
-    # NOTE:  Choose a probability high enough to guarantee at least the
-    # required number of moves, by using a low estimate of the number
-    # of moves per game.  Some experiments show that it's right around
-    # 250 moves/game on average, so 100 is a conservative lower bound.
-    estimate = 100
-    probability = moves / (n * estimate)
-    utils.dbg('Row range: %s - %s; probability %.3f' % (
-        start_row, end_row, probability))
-    ds = _tf_table.parallel_scan_range(start_row, end_row,
-                                       probability=probability,
-                                       columns=[(column_family, column)])
-    # Clip the dataset to the desired number of moves, counting on
-    # the fact that the probability chosen will produce at least the
-    # number of moves required.
-    # TODO: Randomize ordering if it isn't already.
-    ds = ds.take(moves)
+    ds = get_moves_from_games(start, latest_game, moves, shuffle,
+                              column_family, column)
     return ds.map(lambda row_name, s: s)
 
 
+def histogram_move_keys_by_game(sess, ds, batch_size=8*1024):
+    """Given dataset of key names, return histogram of moves/game."""
+    ds = ds.batch(batch_size)
+    iterator = ds.make_initializable_iterator()
+    sess.run(iterator.initializer)
+    get_next = iterator.get_next()
+    h = {}
+    try:
+        while True:
+            bresult = sess.run(
+                tf.reduce_join(get_next,
+                               separator='\n'))
+            result = str(bresult, 'utf-8')
+            for line in result.splitlines():
+                g = line.split('_m_')[0]
+                h[g] = h.get(g, 0) + 1
```

Comment:
consider
`h = defaultdict(int)`
`h[g] += 1`

Reply:
modern python is collections.Counter(iterable)

Reply:
Considered! And after a brief moment of profiling, I'm convinced. `defaultdict` is measurably faster than my idiom. Thanks!